07_leukemia_patient/simulation_patient_cell_lines/workflow.sh

```diff
@@ -0,0 +1,389 @@
+| samtools view -@ "$NTHREADS" \
+| grep -v "CB:Z:-" \
+| grep -v "UB:Z:-" \
+| sort -k28,28 -k27,27 -k1,1 \
+| parallel --pipe --block 10M -j "$NTHREADS" \
+'awk -v seen="" -v current="" -v current_ub_cb="" -v seen_ub_cb="" -v current_ub="" -v seen_ub="" \
+  "BEGIN { OFS=\"\t\" } \
+  { \
+    current = \$1; \
+    current_ub_cb = \$28 \";\" \$27; \
+    current_ub = \$28; \
+    cmd = \"grep -q \" current \" r2.fastq\"; \
+    exit_code = system(cmd); \
+    if (exit_code == 0) { next; } \
+    if (current_ub != seen_ub && current_ub_cb != seen_ub_cb && current != seen) { \
+      print \"@\" current \";\" \$27 \";\" \$28 \"\\n\" \$10 \"\\n+\\n\" \$11; \
+    } \
+    seen = current; \
+    seen_ub_cb = current_ub_cb; \
+    seen_ub = current_ub; \
+  }"' >> r2.fastq
+
+echo "BCR extracted"
+
+samtools view "$STARSOLO_BAM" "chr9:129713016-131887675" -@ "$NTHREADS" \
+| cat header.txt - \
+| samtools view -@ "$NTHREADS" \
+| grep -v "CB:Z:-" \
+| grep -v "UB:Z:-" \
+| sort -k28,28 -k27,27 -k1,1 \
+| parallel --pipe --block 10M -j "$NTHREADS" \
+'awk -v seen="" -v current="" -v current_ub_cb="" -v seen_ub_cb="" -v current_ub="" -v seen_ub="" \
+  "BEGIN { OFS=\"\t\" } \
+  { \
+    current = \$1; \
+    current_ub_cb = \$28 \";\" \$27; \
+    current_ub = \$28; \
+    cmd = \"grep -q \" current \" r2.fastq\"; \
+    exit_code = system(cmd); \
+    if (exit_code == 0) { next; } \
+    if (current_ub != seen_ub && current_ub_cb != seen_ub_cb && current != seen) { \
+      print \"@\" current \";\" \$27 \";\" \$28 \"\\n\" \$10 \"\\n+\\n\" \$11; \
+    } \
+    seen = current; \
+    seen_ub_cb = current_ub_cb; \
+    seen_ub = current_ub; \
+  }"' >> r2.fastq
+
+echo "ABL extracted"
+
+# for unmapped reads: they don't have the gx tag, so the CB and UB will be at a different position (UB (23), CB (22) and 1 (read id)
+
+samtools view -b -f 4 "$STARSOLO_BAM" -@ "$NTHREADS" \
+| samtools view -@ "$NTHREADS" \
+| grep -v "CB:Z:-" \
+| grep -v "UB:Z:-" \
+| sort -k23,23 -k22,22 -k1,1 \
+| parallel --pipe --block 10M -j "$NTHREADS" \
+'awk -v seen="" -v current="" -v current_ub_cb="" -v seen_ub_cb="" -v current_ub="" -v seen_ub="" \
+  "BEGIN { OFS=\"\t\" } \
+  { \
+   current = \$1; \
+   current_ub_cb = \$23 \";\" \$22; \
+   current_ub = \$23; \
+   cmd = \"grep -q \" current \" r2.fastq\"; \
+   exit_code = system(cmd); \
+   if (exit_code == 0) { next; } \
+   if (current_ub != seen_ub && current_ub_cb != seen_ub_cb && current != seen) { \
+     print \"@\" current \";\" \$22 \";\" \$23 \"\\n\" \$10 \"\\n+\\n\" \$11; \
+   } \
+   seen = current; \
+   seen_ub_cb = current_ub_cb; \
+   seen_ub = current_ub; \
+  }"' >> r2.fastq
+
+echo "unmapped extracted"
+
+rm header.txt
+
+## Step 2: additonal CB / UB deduplication
+
+# Additional CB and UB deduplication if reads appended in different steps have same CB and UB
+
+awk 'NR%4==1{print}' r2.fastq | awk '{split ($0, a, /[;]/); print a[2]";"a[3]}' | sort | uniq -d > duplicated.txt
```

Comment:
if you joined the BCR and ABL steps as suggested above this step wouldn't be necessary I guess, because the multialignments for a given read need this read to be aligned (not unmapped)